CONTRIBUTING.md

```diff
@@ -142,3 +125,59 @@ We have some of this in the Reducer code, but generally discourage it.
 - https://github.com/avohq/reasonml-code-style-guide
 - https://cs.brown.edu/courses/cs017/content/docs/reasonml-style.pdf
 - https://github.com/ostera/reason-design-patterns/
+
+# TypeScript style
+
+**Prefer `const` over `let`, never use `var`**
+`var` is deprecated in JS. `let` should only be used for mutable variables.
+
+**Use functional style, avoid classes**
+We use classes for outer-facing APIs, but most of the codebase should use plain immutable objects with functions act on those objects.
```

Comment:
Typescript generics are great! Inheritance through types (instead of classes) is pretty powerful in Typescript too.

There's one performance caveat — generic functions in JS (so both in Typescript and in Rescript) can be slower than monomorphic functions. Lots of details here: https://mrale.ph/blog/2015/01/11/whats-up-with-monomorphism.html.

It doesn't matter for most cases, and we shouldn't avoid generic functions because of this, but it might be important for very hot functions, mostly in the interpreter internals.